identity-account-storage/src/storage/traits.rs

```diff
@@ -27,45 +28,82 @@ mod storage_sub_trait {
-  async fn key_del(&self, did: &IotaDID, location: &KeyLocation) -> Result<()>;
+  /// Deletes the key at `location`.
+  ///
+  /// This operation is idempotent: it does not fail if the key does not exist.
+  async fn key_del(&self, account_id: &AccountId, location: &KeyLocation) -> Result<()>;
 
   /// Signs `data` with the private key at the specified `location`.
-  async fn key_sign(&self, did: &IotaDID, location: &KeyLocation, data: Vec<u8>) -> Result<Signature>;
+  async fn key_sign(&self, account_id: &AccountId, location: &KeyLocation, data: Vec<u8>) -> Result<Signature>;
 
-  /// Returns `true` if a keypair exists at the specified `location`.
-  async fn key_exists(&self, did: &IotaDID, location: &KeyLocation) -> Result<bool>;
+  /// Returns `true` if a key exists at the specified `location`.
+  async fn key_exists(&self, account_id: &AccountId, location: &KeyLocation) -> Result<bool>;
 
-  /// Returns the chain state of the identity specified by `did`.
-  async fn chain_state(&self, did: &IotaDID) -> Result<Option<ChainState>>;
+  /// Returns the chain state of the identity specified by `account_id`.
+  async fn chain_state(&self, account_id: &AccountId) -> Result<Option<ChainState>>;
 
-  /// Set the chain state of the identity specified by `did`.
-  async fn set_chain_state(&self, did: &IotaDID, chain_state: &ChainState) -> Result<()>;
+  /// Set the chain state of the identity specified by `account_id`.
+  async fn set_chain_state(&self, account_id: &AccountId, chain_state: &ChainState) -> Result<()>;
 
-  /// Returns the state of the identity specified by `did`.
-  async fn state(&self, did: &IotaDID) -> Result<Option<IdentityState>>;
+  /// Returns the [`IotaDocument`] of the identity specified by `account_id`.
+  async fn document(&self, account_id: &AccountId) -> Result<Option<IotaDocument>>;
 
-  /// Sets a new state for the identity specified by `did`.
-  async fn set_state(&self, did: &IotaDID, state: &IdentityState) -> Result<()>;
+  /// Sets a new state for the identity specified by `account_id`.
+  async fn set_document(&self, account_id: &AccountId, state: &IotaDocument) -> Result<()>;
 
   /// Removes the keys and any state for the identity specified by `did`.
+  ///
+  /// This operation is idempotent: it does not fail if the given `did` does not exist.
   async fn purge(&self, did: &IotaDID) -> Result<()>;
+
+  /// Adds the `did -> account_id` mapping to the index.
+  ///
+  /// Note that his operation needs to be synchronized globally for a given storage instance,
+  /// in order to prevent race conditions when inserting index entries concurrently.
+  async fn index_set(&self, did: IotaDID, account_id: AccountId) -> Result<()>;
+
+  /// Looks up the [`AccountId`] of the given `did`, returning `None` if no mapping exists.
+  async fn index_get(&self, did: &IotaDID) -> Result<Option<AccountId>>;
+
+  /// Retrieves the list of [`IotaDID`]s stored in the index.
+  async fn index(&self) -> Result<Vec<IotaDID>>;
```

Comment:
Shouldn't this return a map of `IotaDID`s to `AccountId`s? Maybe `index_keys` would make more sense?

Reply:
Should be handled with #757.

Reply:
Not really sure how key deletion affects the `AccountId` index and `index_keys`?

Reply:
Ah, sorry! I meant to post this on the comment above. I will address the `index_keys` comment.

Reply:
To address the actual comment: I felt like a vector is easier to construct for storage and it should also be faster than copying dids and account ids from storage into a hashmap. I don't think we ever need to return a `HashMap<IotaDID, AccountId>` to users, because `AccountId` is a private type. Thus, a list of DIDs should be good enough? For internal lookups, `index_get` should be used instead of (potentially wastefully) creating the hash map and using that to do the lookup.
For both our storage implementations returning the entire hash map would be trivial, but that's due to a rather wasteful implementation (specifically in stronghold), but that doesn't need to hold true for other implementations.